docs/design/endpoints.md

```diff
@@ -0,0 +1,292 @@
+taken in order to determine the endpoint that calls should be made to. This language is provided inline with the model
+as the `smithy.rules#endpointRuleSet` service trait. The schema for this trait will be described as part of Smithy's
+extended documentation in the future.
+
+The smithy-kotlin repository implements the core code-generation for endpoint providers based on rule sets defined this
+way, as well as the middleware that sources the parameters for the resolver and calls it on each request.
+
+## Core Concepts
+### Endpoint
+The `Endpoint` is the result of calling an endpoint provider implementation at runtime to determine what endpoint to use
+when making service calls. The type is public and shared across all service clients, and all provider implementations
+MUST return values of this type. All values of the endpoint are provided by the ruleset.
+
+The (annotated) type is provided here:
+
+```kotlin
+public data class Endpoint @InternalApi constructor(
+    /**
+     * The actual URL against which service operations are called.
+     */
+    public val uri: Url,
+
+    /**
+     * Any additional HTTP headers to send on calls to this endpoint.
+     */
+    public val headers: Headers = Headers.Empty,
+
+    /**
+     * SDK-specific attributes tied to this endpoint.
+     * 
+     * The values within the set of attributes are unstable. Additionally, custom provider implementations cannot set
+     * these values as part of endpoints they resolve to (hence the internal annotation). They can only be specified
+     * in the Smithy ruleset and returned as part of generated code therein.
+     * 
+     * SDK authors are encouraged to define strongly-typed extensions for values relevant to their SDK's operation.
+     */
+    @InternalApi
+    public val attributes: Attributes = Attributes(),
+)
+```
+
+### Endpoint provider
+
+Providers house the core logic for determining the endpoint to use **on a per-request basis**. Providers are generated
+as both a service-specific interface (to be tied to that service's specific provider parameters type and allow for
+custom implementations) and a default provider that implements the logic defined in the model.
+
+```kotlin
+public fun interface EndpointProvider {
+    public suspend fun resolveEndpoint(params: ServiceEndpointParams): Endpoint
+}
+```
+Client config will include a field for the resolver unique to that service, and SDK operations will use the generated
+default unless specified otherwise by the caller.
+
+### Provider library
+The endpoint rulesets as provided by Smithy require some "standard-library"-type functions that support or drive the
+evaluation of the rules.
+
+Some of these are trivial (eg. "is a value non-null") and are generated inline as part of the
+ruleset evaluation. The more complex functions (eg. URL parsing) are hosted in the runtime at
+`package aws.smithy.kotlin.runtime.http.endpoints.functions`. Runtime-hosted standard library functions are annotated as
```

Comment:
Tangent tot his design, something I've wondered is whether we should pull endpoints out of `http` into it's own module (or core depending). The only hiccup here (and likely why I placed it here in the first place) is that the type relies on `Url`. Something to think about whether we want to tie endpoints to http though.